src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -57,127 +58,34 @@ public abstract class EnumeratingWhitelist extends Whitelist {
 
     protected abstract List<FieldSignature> staticFieldSignatures();
 
-    ConcurrentHashMap<String, Boolean> permittedCache = new ConcurrentHashMap<>();  // Not private to facilitate testing
-
-    @SafeVarargs
-    private final void cacheSignatureList(List<Signature> ...sigs) {
-        for (List<Signature> list : sigs) {
-            for (Signature s : list) {
-                if (!s.isWildcard()) { // Cache entries for wildcard signatures will never be accessed and just waste space
-                    permittedCache.put(s.toString(), Boolean.TRUE);
-                }
-            }
-        }
-    }
-
-    /** Prepopulates the "permitted" cache, resetting if populated already.  Should be called when method signatures change or after initialization. */
-    final void precache() {
-        if (!permittedCache.isEmpty()) {
-            this.permittedCache.clear();  // No sense calling clearCache
-        }
-        cacheSignatureList((List)methodSignatures(), (List)(newSignatures()),
-                           (List)(staticMethodSignatures()), (List)(fieldSignatures()),
-                           (List)(staticFieldSignatures()));
-    }
-
-    /** Frees up nearly all memory used for the cache.  MUST BE CALLED if you change the result of the xxSignatures() methods. */
-    final void clearCache() {
-        this.permittedCache.clear();
-        this.permittedCache = new ConcurrentHashMap<>();
-    }
+    private final Cache<Method, Boolean> methodCache = Caffeine.newBuilder().weakKeys().build();
+    private final Cache<Constructor<?>, Boolean> constructorCache = Caffeine.newBuilder().weakKeys().build();
+    private final Cache<Method, Boolean> staticMethodCache = Caffeine.newBuilder().weakKeys().build();
+    private final Cache<Field, Boolean> fieldCache = Caffeine.newBuilder().weakKeys().build();
+    private final Cache<Field, Boolean> staticFieldCache = Caffeine.newBuilder().weakKeys().build();
 
     @Override public final boolean permitsMethod(@NonNull Method method, @NonNull Object receiver, @NonNull Object[] args) {
-        String key = canonicalMethodSig(method);
```

Comment:
Note that this will no longer cache results when there is a distinct `Method` with the same signature, which could happen when there is a new copy of the `Class` from a distinct `ClassLoader`. However https://github.com/jenkinsci/workflow-cps-plugin/blob/3a59e40fb41d2774421ce3d7c00480e5917edb76/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/GroovyClassLoaderWhitelist.java#L45 consults its `delegate` https://github.com/jenkinsci/workflow-cps-plugin/blob/3a59e40fb41d2774421ce3d7c00480e5917edb76/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsGroovyShell.java#L181 last so I do not think that would normally happen. Once the system warms up, the normal path through ought to be fast. Would be useful to have some JMH benchmarks if they can run via `RealJenkinsRule`.